### ISSUE TITLE
Add waffle flag to turn submissions on or off

### ISSUE THREAD

**Issue by diox** [mozilla/addons#1830]
We want to add something admins could easily toggle to turn submissions on or off (devhub + old and new APIs) in case of emergencies - we already have solutions, but they require ops, and we want something more granular.

Ideally, this would be some kind of custom Waffle [flag](https://waffle.readthedocs.io/en/stable/types/flag.html), enabling us to have an option allowing submissions of add-ons matching some criteria (like recommended) to still go through if we so wish.

It should also return a custom error message and HTTP status code that clearly states why submissions are turned off temporarily (also something that waffle `Flag` model supports, through the `note` attribute - it would be English only, but that should be acceptable for this feature). We should make sure that error message is clearly visible in devhub and web-ext.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/AMOENG-116)

**Comment by diox** [comment]
Need scope clarification on "have an option allowing submissions of add-ons matching some criteria (like recommended) to still go through if we so wish."

**Comment by diox** [comment]
@wagnerand what do you think about https://github.com/mozilla/addons/issues/1830#issuecomment-2338213445 ? If it's not needed, it obviously makes our life much easier. If it is, we need the exact criteria/mechanism wanted for bypass (which could be optional, but we still need to know what it is to build it).

**Comment by wagnerand** [comment]
I could see an override option for admins to submit a new version for any add-on, in case there is a critical request from a developer (similar to admins being allowed to override a failed validation). I am not sure we should have exceptions for certain sets of add-ons.

@abyrne-moz might want to weigh in on this as well.

**Comment by abyrne-moz** [comment]
Can you provide an example scenario of when we might need to use this?  "In case of emergencies" isn't that clear to me, and I assume an emergency lasts a short period of time and therefore limits the need for an override option.

**Comment by diox** [comment]
A bypass for admins is trivial for us to implement through a waffle `Flag` (we'd essentially be adding a `Flag` that defaults to `everyone`, and when we want to turn submissions off, we'd flip that `everyone` bit to `False`, and we could still have `superusers` left at `True` to allow admins to submit stuff)

**Comment by diox** [comment]
@wagnerand @abyrne-moz 

We need copy for this feature. Currently in [the PR](https://github.com/mozilla/addons-server/pull/22729) we have:
- Submissions Unavailable (page title)
- Submissions are not currently available (page header)
- Submissions have been temporarily disabled. Please check back later. (default page contents)

I'm thinking "Submissions" is too generic here. Could you help come up with something?

**Comment by abyrne-moz** [comment]
Would "Add-on uploads" be more appropriate? 

-     "Add-on uploads have been temporarily disabled. Please check back later". That's the default if no message is set in the waffle Flag.
On top of that, we have a special error page if somehow the developers bypasses disabled buttons or has unlucky timing and submits an add-on right as the submissions are being disabled. That page has a title and a header:
-     "Add-on uploads temporarily disabled" (page title)
-     "Add-on uploads are temporarily unavailable" (page header)

**Comment by diox** [comment]
Yeah I think that's good. cc @chrstinalin ^

**Comment by ioanarusiczki** [comment]
I did not change anything to the `enable-submissions` only switched between yes/no/unknown (so I left  Superusers checked)

With switch off (no/unknown)
- on dev hub add-on uploads are not available for a regular user or an admin (and I think I should be able to do that as an admin)
- with web-ext I get different results. A first version submitted as a regular user or admin is not possible, I get an error message
"Service Unavailable". But I can submit new versions (as a developer or admin) which I don't think is expected. 

I must continue testing around but I thought I should leave an update so far (I hope I am using the correct settings for the flag)

**Comment by diox** [comment]
We don't have the decorator on `FileUploadViewSet.create()` and we probably should, but that shouldn't be necessary to prevent `web-ext` submissions, since we still need to use the `upload`, so we're probably missing it somewhere else ? Need to investigate.

The admin not being able to bypass the flag if `superusers` is `True` is also weird.

**Comment by ioanarusiczki** [comment]
I'll add my latest test results:

- I tried new addon uploads and new versions from Postman with a regular and then an admin account with "*" permission.  Here I hit the 503 every time (I could not make new version uploads with those I tried for web-ext 633391 or 633393)

![503](https://github.com/user-attachments/assets/d5765cee-e0a3-4996-896e-f5d12f8e6d26)

- from web-ext I could still continue to make new version uploads (I've tested this from AMO dev)
the addon from the admin account https://reviewers.addons-dev.allizom.org/en-US/reviewers/review-unlisted/633391
the addon with a regular user https://reviewers.addons-dev.allizom.org/en-US/reviewers/review/633393

- dev hub looks good but I'll mention the theme wizard. If I upload an image the button becomes enabled yet the submission is not going to finish. 

![themewizard zip](https://github.com/user-attachments/assets/c2d5e860-05b0-43a2-b698-74c345288a97)

**Comment by chrstinalin** [comment]
> We don't have the decorator on `FileUploadViewSet.create()` and we probably should, but that shouldn't be necessary to prevent `web-ext` submissions, since we still need to use the `upload`, so we're probably missing it somewhere else ? Need to investigate.

I checked the `web-ext` repo, and it looks like they [directly](https://github.com/mozilla/web-ext/blob/420aa78db72f6c8d10f7ae6cfa8a61be7cf43820/src/util/submit-addon.js#L127) use `/addons/upload` in the submission process, i.e `FileUploadViewSet.create()`, which is possibly why it's behaving differently. I did have a decorator on that endpoint originally, but removed it for a reason that's... lost on me now. 🙃


> The admin not being able to bypass the flag if `superusers` is `True` is also weird.
This I'm stumped on. Does `flag_is_active` not handle this? I'll look into it some more.

**Comment by diox** [comment]
> This I'm stumped on. Does `flag_is_active` not handle this? I'll look into it some more.

Also stumped on this. `<flag>.is_active(request)` should definitely handle this. There is also a `<flag>.is_active_for_user(user)` method and when creating a flag with `everyone=False` and `superusers=True` it seemed to work fine for me. Maybe some caching involved? Very weird.

**Comment by chrstinalin** [comment]
Submissions seem to disable correctly for me locally when `everyone: None` and `superusers: True`.

> on dev hub add-on uploads are not available for a regular user or an admin (and I think I should be able to do that as an admin)

@ioanarusiczki could I ask what the exact steps you took here were?

**Comment by chrstinalin** [comment]
(Figured it out over Slack -- Everyone was set to `No` and overrode all other permissions).

**Comment by ioanarusiczki** [comment]
Yes, so what I've tested is an expected result for admins when the flag has "No" instead of "Unknown" set.

But with state unknown it's different:

An admin can upload new addons or versions from dev hub on his account or someone else's behalf. Using directly the endpoints it's different. It will throw a 503 Service Unavailable. Admins cannot submit using the endpoints versions when they're not owners of those extensions anyway so this might not be a problem. I don't know if it's expected to be able to upload their own extensions though. 

An admin reviewer cannot submit addons from dev hub. -> confirms that only Admins with * can do this.

A regular user cannot submit extensions using dev hub or the endpoints. -> expected

If a regular user is added into the flag as an exception, uploads can be done only from dev hub. With endpoints I'll get a 503. Maybe that's also expected? 

With flag set as currently on stage (Yes for Everyone + Superusers checked) -> submissions are behaving normally. Everyone can make uploads.

**Comment by diox** [comment]
> If a regular user is added into the flag as an exception, uploads can be done only from dev hub. With endpoints I'll get a 503. Maybe that's also expected?

Please file a follow-up for that. It probably has to do with how the authentication works in the API.

**Comment by ioanarusiczki** [comment]
@diox 

And when The flag has the Unknown state with Superusers checked.  Should an admin also be allowed to make his own uploads from the api ? (I know I made a complicated explanation above)

> An admin can upload new addons or versions from dev hub on his account or someone else's behalf. Using directly the endpoints it's different. It will throw a 503 Service Unavailable. Admins cannot submit using the endpoints versions when they're not owners of those extensions anyway so this might not be a problem. I don't know if it's expected to be able to upload their own extensions though.

**Comment by diox** [comment]
They should be able to, and that's probably the same problem.

**Comment by ioanarusiczki** [comment]
So it's not a surprise that https://github.com/mozilla/addons/issues/15096 is reproducible with web-ext too.

I verified on -dev the web-ext submissions. New versions cannot be submitted while switch had Unknown state with a regular user or an admin.
